### PR TITLE
Add __builtins__ to globals()

### PR DESCRIPTION
**Pull request by youknowone**
This feature allows to expand `__builtins__` for program-global

### CODE REVIEW

File: vm/src/pyobject.rs [hunk 1]
```diff
@@ -643,8 +643,13 @@ impl PyContext {
         objtype::new(self.type_type(), name, vec![base], PyAttributes::new()).unwrap()
     }
 
-    pub fn new_scope(&self) -> Scope {
-        Scope::new(None, self.new_dict())
+    pub fn new_scope(&self, vm: &VirtualMachine) -> Scope {
+        let scope = Scope::new(None, self.new_dict());
+
+        let globals = scope.globals.clone();
+        let _ = globals.set_item("__builtins__", vm.builtins.clone(), vm);
```
Comment (windelbouwman) [inline vm/src/pyobject.rs]: I'm unsure whether this is the right place to set the item. Anyways, the function name should be changed, since the function now no longer creates a new scope. I would suggest to leave the original function in place, and create a second function which creates a scope (by calling the original function) and adds the builtin module. Then we can use this new function when we require a scope with builtins.

In other words: do we always require builtins in the scope?

Comment (windelbouwman) [inline vm/src/pyobject.rs]: You can remove the `let _ =` part here. Underscore is the unused value. In case set_item returns an error, it should be handled by the `?` operator, or by unwrap, such that we panic when this fails. My preference is to handle the error with the `?` operator.

File: benchmarks/bench.rs [hunk 2]
```diff
@@ -99,7 +99,7 @@ fn bench_rustpy_nbody(b: &mut test::Bencher) {
     };
 
     b.iter(|| {
-        let scope = vm.ctx.new_scope();
+        let scope = vm.ctx.new_scope(&vm);
```
Comment (cthulahoops) [inline benchmarks/bench.rs]: I don't like the method on context taking a ```VirtualMachine```. Given that this now needs vm then I think the method should move to the ```VirtualMachine```.

File: vm/src/pyobject.rs (hunk 1 above)
Comment (cthulahoops) [reply vm/src/pyobject.rs]: I think ```__builtins__``` get added in every case (except presumably during the creation of builtins itself). Even if we pass a dictionary to as globals, ```__builtins__``` gets added:

```
>>> exec('assert "__builtins__" in globals()', dict())
```

It doesn't get overwritten if already there:

```
exec('assert __builtins__ == 7', {'__builtins__': 7})
```

And the ```__builtins__``` provided is used for resolution of builtin names:

```
>>> exec('print()', {'__builtins__': {}})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
NameError: name 'print' is not defined
```

And it doesn't use set_item to add it to the dictionary. (If you use a custom class for globals you don't get to intercept the addition of builtins.

There's also something odd that ```__builtins__``` is sometimes a module and sometimes a dict.

I'm not sure what the best implementation is. I lean towards not changing ```new_scope``` and doing the insert in the caller (because some callers will have to do special things to add it to globals that lack it.

**Comment by youknowone** [comment]
I have no idea about exec examples.
But I created `Scope::init_builtins()` to do whatever required builtins stuff to the scope in the future.

File: wasm/lib/src/vm_class.rs [hunk 3]
```diff
@@ -27,6 +27,7 @@ impl StoredVirtualMachine {
     fn new(id: String, inject_browser_module: bool) -> StoredVirtualMachine {
         let mut vm = VirtualMachine::new();
         let scope = vm.ctx.new_scope();
+        let _ = scope.init_builtins(&vm);
```
Comment (windelbouwman) [inline wasm/lib/src/vm_class.rs]: Instead of ignoring the result, use `.unwrap()` instead. Then the error will be noticed earlier. The init_builtins function is to pass always, so this should not be a problem.

**Comment by youknowone** [comment]
Fixed about `let _ =`

**Comment by youknowone** [comment]
@windelbouwman @cthulahoops  I revised second commit. I am sorry that I made change after review.

Now, `__builtins__` is dict when exec() got dict variable; otherwise module. I think this is now correct rule for `globals()`.
Name resolving is a little bit different. For now, we lookup builtins module directly but not lookup `__builtins__`.

File: vm/src/builtins.rs [hunk 4]
```diff
@@ -218,18 +218,30 @@ fn make_scope(
...
         None => None,
     };
-
     let current_scope = vm.current_scope();
     let globals = match globals {
-        Some(dict) => dict.clone().downcast().unwrap(),
+        Some(dict) => {
+            let dict: PyDictRef = dict.clone().downcast().unwrap();
+            if !dict.contains_key("__builtins__", vm) {
+                let builtins_dict = match &vm.builtins.dict {
```
Comment (windelbouwman) [inline vm/src/builtins.rs]: Please change this match statement into an `.unwrap()` call. This will panic as well:
```rust
let builtins_dict = &vm.builtins.dict.unwrap().as_object();
```

File: vm/src/frame.rs [hunk 5]
```diff
@@ -90,7 +90,13 @@ impl fmt::Debug for Scope {
 }
 
 impl Scope {
-    pub fn new(locals: Option<PyDictRef>, globals: PyDictRef) -> Scope {
+    pub fn new(locals: Option<PyDictRef>, globals: PyDictRef, vm: &VirtualMachine) -> Scope {
+        if !globals.contains_key("__builtins__", vm) {
```
Comment (windelbouwman) [inline vm/src/frame.rs]: I would not set the builtins in this new function, but in the calling sites of this new function, as @cthulahoops suggested.

File: vm/src/frame.rs [hunk 6]
```diff
@@ -90,7 +90,13 @@ impl fmt::Debug for Scope {
 }
 
 impl Scope {
-    pub fn new(locals: Option<PyDictRef>, globals: PyDictRef) -> Scope {
+    pub fn new(locals: Option<PyDictRef>, globals: PyDictRef, vm: &VirtualMachine) -> Scope {
```
Comment (windelbouwman) [inline vm/src/frame.rs]: This function can be left as is, because we will add builtins on other locations when required.

File: vm/src/vm.rs [hunk 7]
```diff
@@ -250,6 +250,10 @@ impl VirtualMachine {
         syntax_error
     }
 
+    pub fn new_scope(&self) -> Scope {
+        Scope::new(None, self.ctx.new_dict(), self)
```
Comment (windelbouwman) [inline vm/src/vm.rs]: Please remove the last `self` argument.

File: vm/src/frame.rs (hunk 5 above)
Comment (youknowone) [reply vm/src/frame.rs]: I thought @cthulahoops said `__builtins__` is added in every case.

If we still need to keep the `new` clean, I will add `Scope::with_builtins`. Is this ok?

File: vm/src/builtins.rs (hunk 4 above)
Comment (youknowone) [reply vm/src/builtins.rs]: When I tried it, I got error like this:
```rust
error[E0507]: cannot move out of an `Rc`
   --> vm/src/builtins.rs:226:38
    |
226 |                 let builtins_dict = &vm.builtins.dict.unwrap().as_object();
    |                                      ^^^^^^^^^^^^^^^^ cannot move out of an `Rc`
```

**Comment by youknowone** [comment]
I reverted `Scope::new()` as before and added a new method `Scope::with_builtins()`

File: vm/src/builtins.rs (hunk 4 above)
Comment (windelbouwman) [reply vm/src/builtins.rs]: Maybe this works better:
```rust
let builtins_dict = vm.builtins.dict.as_ref().unwrap().as_object();
```

File: tests/snippets/builtins_module.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+exec('assert "__builtins__" in globals()', dict())
+exec('assert __builtins__ == 7', {'__builtins__': 7})
+exec('assert not isinstance(__builtins__, dict)')
+exec('assert isinstance(__builtins__, dict)', {})
+
+namespace = {}
+exec('', namespace)
+assert namespace['__builtins__'] == __builtins__.__dict__
+
+# with assertRaises(NameError):
```
Comment (windelbouwman) [inline tests/snippets/builtins_module.py]: What is the issue with this test? Can this test be enabled?

File: vm/src/vm.rs [hunk 9]
```diff
@@ -250,6 +250,10 @@ impl VirtualMachine {
         syntax_error
     }
 
+    pub fn new_scope(&self) -> Scope {
```
Comment (windelbouwman) [inline vm/src/vm.rs]: Please rename this function to reflect that it creates a scope and adds builtins. Something like: `new_scope_with_builtins`.

File: tests/snippets/builtins_module.py (hunk 8 above)
Comment (youknowone) [reply tests/snippets/builtins_module.py]: In RustPython, `exec('print()')` always succeed without builtins injection (original state).
Because this patch didn't touch anything about name resolving, this code still works - so the test is failed.

File: vm/src/vm.rs (hunk 9 above)
Comment (youknowone) [reply vm/src/vm.rs]: Fixed it

File: vm/src/builtins.rs (hunk 4 above)
Comment (youknowone) [reply vm/src/builtins.rs]: Thanks! this code works well

**Comment by windelbouwman** [comment]
@youknowone I will merge this now. Thank you for your patience with this pull request, I think it is now in good shape.
